### ISSUE TITLE
Unit tests are failing for Terry

### ISSUE THREAD

**Issue by edreamleo** [leo-editor/leo-editor#512]
Terry says: Here are the fails I get from d1608d30. It would really help if the fail count could be assumed to be zero on pulls from master.

```
Trying to create a QVariant instance of QMetaType::Void type, an
invalid QVariant will be constructed instead

[this is repeated many many times]

======================================================================
ERROR: runTest (leo.core.leoTest.GeneralTestCase)
@test rST import test

----------------------------------------------------------------------
Traceback (most recent call last):
  File "/mnt/usr1/usr1/home/tbrown/t/Package/leo/git/leo-editor/leo/core/leoTest.py", line 211, in runTest
    builtins.execfile(scriptFile, d)
  File "/home/tbrown/.leo/scriptFile.py", line 74, in <module>
    assert root.h.startswith('@@'), root.h
  File "/mnt/usr1/usr1/home/tbrown/t/Package/leo/git/leo-editor/leo/core/leoNodes.py", line 1635, in __get_h
    return p.headString()
  File "/mnt/usr1/usr1/home/tbrown/t/Package/leo/git/leo-editor/leo/core/leoNodes.py", line 672, in headString
    return self.v.headString()
AttributeError: 'NoneType' object has no attribute 'headString'

======================================================================
ERROR: runTest (leo.core.leoTest.GeneralTestCase)
@test rST import test: simple

----------------------------------------------------------------------
Traceback (most recent call last):
  File "/mnt/usr1/usr1/home/tbrown/t/Package/leo/git/leo-editor/leo/core/leoTest.py", line 211, in runTest
    builtins.execfile(scriptFile, d)
  File "/home/tbrown/.leo/scriptFile.py", line 22, in <module>
    assert root.h.startswith('@@'), root.h
  File "/mnt/usr1/usr1/home/tbrown/t/Package/leo/git/leo-editor/leo/core/leoNodes.py", line 1635, in __get_h
    return p.headString()
  File "/mnt/usr1/usr1/home/tbrown/t/Package/leo/git/leo-editor/leo/core/leoNodes.py", line 672, in headString
    return self.v.headString()
AttributeError: 'NoneType' object has no attribute 'headString'

======================================================================
ERROR: runTest (leo.core.leoTest.GeneralTestCase)
@test rST import test: no double-underlines

----------------------------------------------------------------------
Traceback (most recent call last):
  File "/mnt/usr1/usr1/home/tbrown/t/Package/leo/git/leo-editor/leo/core/leoTest.py", line 211, in runTest
    builtins.execfile(scriptFile, d)
  File "/home/tbrown/.leo/scriptFile.py", line 73, in <module>
    assert root.h.startswith('@@'), root.h
  File "/mnt/usr1/usr1/home/tbrown/t/Package/leo/git/leo-editor/leo/core/leoNodes.py", line 1635, in __get_h
    return p.headString()
  File "/mnt/usr1/usr1/home/tbrown/t/Package/leo/git/leo-editor/leo/core/leoNodes.py", line 672, in headString
    return self.v.headString()
AttributeError: 'NoneType' object has no attribute 'headString'

======================================================================
ERROR: runTest (leo.core.leoTest.GeneralTestCase)
@test rST import test: long underlines

----------------------------------------------------------------------
Traceback (most recent call last):
  File "/mnt/usr1/usr1/home/tbrown/t/Package/leo/git/leo-editor/leo/core/leoTest.py", line 211, in runTest
    builtins.execfile(scriptFile, d)
  File "/home/tbrown/.leo/scriptFile.py", line 34, in <module>
    assert root.h.startswith('@@'), root.h
  File "/mnt/usr1/usr1/home/tbrown/t/Package/leo/git/leo-editor/leo/core/leoNodes.py", line 1635, in __get_h
    return p.headString()
  File "/mnt/usr1/usr1/home/tbrown/t/Package/leo/git/leo-editor/leo/core/leoNodes.py", line 672, in headString
    return self.v.headString()
AttributeError: 'NoneType' object has no attribute 'headString'

======================================================================
ERROR: runTest (leo.core.leoTest.GeneralTestCase)
@test rST import test: long overlines

----------------------------------------------------------------------
Traceback (most recent call last):
  File "/mnt/usr1/usr1/home/tbrown/t/Package/leo/git/leo-editor/leo/core/leoTest.py", line 211, in runTest
    builtins.execfile(scriptFile, d)
  File "/home/tbrown/.leo/scriptFile.py", line 35, in <module>
    assert root.h.startswith('@@'), root.h
  File "/mnt/usr1/usr1/home/tbrown/t/Package/leo/git/leo-editor/leo/core/leoNodes.py", line 1635, in __get_h
    return p.headString()
  File "/mnt/usr1/usr1/home/tbrown/t/Package/leo/git/leo-editor/leo/core/leoNodes.py", line 672, in headString
    return self.v.headString()
AttributeError: 'NoneType' object has no attribute 'headString'

======================================================================
ERROR: runTest (leo.core.leoTest.GeneralTestCase)
@test rST import test: trailing whitespace

----------------------------------------------------------------------
Traceback (most recent call last):
  File "/mnt/usr1/usr1/home/tbrown/t/Package/leo/git/leo-editor/leo/core/leoTest.py", line 211, in runTest
    builtins.execfile(scriptFile, d)
  File "/home/tbrown/.leo/scriptFile.py", line 22, in <module>
    assert root.h.startswith('@@'), root.h
  File "/mnt/usr1/usr1/home/tbrown/t/Package/leo/git/leo-editor/leo/core/leoNodes.py", line 1635, in __get_h
    return p.headString()
  File "/mnt/usr1/usr1/home/tbrown/t/Package/leo/git/leo-editor/leo/core/leoNodes.py", line 672, in headString
    return self.v.headString()
AttributeError: 'NoneType' object has no attribute 'headString'

======================================================================
ERROR: runTest (leo.core.leoTest.GeneralTestCase)
@test leo_rst

----------------------------------------------------------------------
Traceback (most recent call last):
  File "/mnt/usr1/usr1/home/tbrown/t/Package/leo/git/leo-editor/leo/core/leoTest.py", line 211, in runTest
    builtins.execfile(scriptFile, d)
  File "/home/tbrown/.leo/scriptFile.py", line 48, in <module>
    assert root.h.startswith('@@'), root.h
  File "/mnt/usr1/usr1/home/tbrown/t/Package/leo/git/leo-editor/leo/core/leoNodes.py", line 1635, in __get_h
    return p.headString()
  File "/mnt/usr1/usr1/home/tbrown/t/Package/leo/git/leo-editor/leo/core/leoNodes.py", line 672, in headString
    return self.v.headString()
AttributeError: 'NoneType' object has no attribute 'headString'

======================================================================
ERROR: runTest (leo.core.leoTest.GeneralTestCase)
@test x.makeShadowDirectory

----------------------------------------------------------------------
Traceback (most recent call last):
  File "/mnt/usr1/usr1/home/tbrown/t/Package/leo/git/leo-editor/leo/core/leoTest.py", line 211, in runTest
    builtins.execfile(scriptFile, d)
  File "/home/tbrown/.leo/scriptFile.py", line 31, in <module>
    deleteShadowDir(shadow_dir)
  File "/home/tbrown/.leo/scriptFile.py", line 20, in deleteShadowDir
    os.rmdir(shadow_dir)
OSError: [Errno 39] Directory not empty: '/home/tbrown/.leo/shadow'

======================================================================
FAIL: runTest (leo.core.leoTest.GeneralTestCase)
@test all commands have an event arg

----------------------------------------------------------------------
Traceback (most recent call last):
  File "/mnt/usr1/usr1/home/tbrown/t/Package/leo/git/leo-editor/leo/core/leoTest.py", line 211, in runTest
    builtins.execfile(scriptFile, d)
  File "/home/tbrown/.leo/scriptFile.py", line 26, in <module>
    assert arg0 in expected or arg1 in expected, message
AssertionError:
no event arg for command quickmove_bookmark_to_first_child, func: <lambda>
args: ArgSpec(args=['e'], varargs=None, keywords=None, defaults=None)

======================================================================
FAIL: runTest (leo.core.leoTest.GeneralTestCase)
@test add/delete html comments

----------------------------------------------------------------------
Traceback (most recent call last):
  File "/mnt/usr1/usr1/home/tbrown/t/Package/leo/git/leo-editor/leo/core/leoTest.py", line 211, in runTest
    builtins.execfile(scriptFile, d)
  File "/home/tbrown/.leo/scriptFile.py", line 27, in <module>
    assert p.b == s,'fail5: s\n%s\nresult\n%s' % (repr(s),repr(p.b))
AssertionError: fail5: s
u'@language html\n<html>\n    text \n</html>\n'
result
u'@language html\n<html>\n   text \n</html>\n'

======================================================================
FAIL: runTest (leo.core.leoTest.GeneralTestCase)
@test helpForbindings

----------------------------------------------------------------------
Traceback (most recent call last):
  File "/mnt/usr1/usr1/home/tbrown/t/Package/leo/git/leo-editor/leo/core/leoTest.py", line 211, in runTest
    builtins.execfile(scriptFile, d)
  File "/home/tbrown/.leo/scriptFile.py", line 7, in <module>
    c.helpCommands.helpForBindings()
  File "/mnt/usr1/usr1/home/tbrown/t/Package/leo/git/leo-editor/leo/commands/helpCommands.py", line 326, in helpForBindings
    self.c.putHelpFor(s)
  File "/mnt/usr1/usr1/home/tbrown/t/Package/leo/git/leo-editor/leo/core/leoCommands.py", line 7020, in putHelpFor
    g.app.gui.put_help(c, s, short_title)
  File "/mnt/usr1/usr1/home/tbrown/t/Package/leo/git/leo-editor/leo/plugins/qt_gui.py", line 1029, in put_help
    assert vr # For unit testing.
AssertionError

======================================================================
FAIL: runTest (leo.core.leoTest.GeneralTestCase)
@test helpForFindCommands

----------------------------------------------------------------------
Traceback (most recent call last):
  File "/mnt/usr1/usr1/home/tbrown/t/Package/leo/git/leo-editor/leo/core/leoTest.py", line 211, in runTest
    builtins.execfile(scriptFile, d)
  File "/home/tbrown/.leo/scriptFile.py", line 7, in <module>
    c.helpCommands.helpForFindCommands()
  File "/mnt/usr1/usr1/home/tbrown/t/Package/leo/git/leo-editor/leo/commands/helpCommands.py", line 756, in helpForFindCommands
    self.c.putHelpFor(s)
  File "/mnt/usr1/usr1/home/tbrown/t/Package/leo/git/leo-editor/leo/core/leoCommands.py", line 7020, in putHelpFor
    g.app.gui.put_help(c, s, short_title)
  File "/mnt/usr1/usr1/home/tbrown/t/Package/leo/git/leo-editor/leo/plugins/qt_gui.py", line 1029, in put_help
    assert vr # For unit testing.
AssertionError

======================================================================
FAIL: runTest (leo.core.leoTest.GeneralTestCase)
@test helpForMinibuffer

----------------------------------------------------------------------
Traceback (most recent call last):
  File "/mnt/usr1/usr1/home/tbrown/t/Package/leo/git/leo-editor/leo/core/leoTest.py", line 211, in runTest
    builtins.execfile(scriptFile, d)
  File "/home/tbrown/.leo/scriptFile.py", line 6, in <module>
    c.helpCommands.helpForMinibuffer()
  File "/mnt/usr1/usr1/home/tbrown/t/Package/leo/git/leo-editor/leo/commands/helpCommands.py", line 815, in helpForMinibuffer
    c.putHelpFor(s)
  File "/mnt/usr1/usr1/home/tbrown/t/Package/leo/git/leo-editor/leo/core/leoCommands.py", line 7020, in putHelpFor
    g.app.gui.put_help(c, s, short_title)
  File "/mnt/usr1/usr1/home/tbrown/t/Package/leo/git/leo-editor/leo/plugins/qt_gui.py", line 1029, in put_help
    assert vr # For unit testing.
AssertionError

======================================================================
FAIL: runTest (leo.core.leoTest.EditBodyTestCase)
EditBodyTestCase: addComments
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/mnt/usr1/usr1/home/tbrown/t/Package/leo/git/leo-editor/leo/core/leoTest.py", line 105, in runTest
    self.editBody()
  File "/mnt/usr1/usr1/home/tbrown/t/Package/leo/git/leo-editor/leo/core/leoTest.py", line 85, in editBody
    compareHeadlines=False), '%s: before undo1' % commandName
AssertionError: addComments: before undo1

----------------------------------------------------------------------
Ran 904 tests in 108.952s

FAILED (failures=6, errors=8, skipped=28)
tb:la: ~/t/Package/leo/git/leo-editor
leo-editorlocalhost:11.0>
```

**Comment by edreamleo** [comment]
Rev 52c083 skips rst-related unit tests if docutils does not exist.

@tbnorth please rerun tests and report any failures.

**Comment by edreamleo** [comment]
@vitalije What happens when your run unit tests now, either in master or in your branch?

**Comment by edreamleo** [comment]
I am going to close this issue.  Anyone is free to re-open this issue if any unit tests fails when run in the master branch.

**Comment by vitalije** [comment]
@edreamleo 
I just pulled latest commit and rerun unitTest.leo. I still have 6 Errors  while executing tests.
I have installed docutils module.

**Comment by vitalije** [comment]
`
reading settings in /media/vitalije/home11/vitalije/programi/leo-editor/trunk-git/leo/config/leoSettings.leo
reading settings in /media/vitalije/home11/vitalije/programi/leo-editor/trunk-git/leo/config/myLeoSettings.leo
reading settings in /media/vitalije/home11/vitalije/programi/leo-editor/trunk-git/leo/test/unitTest.leo
reading: @file unitTestStartup.txt
reading: @file activeUnitTests.txt

** isPython3: False
Leo 5.5, build 20170702061749, Sun Jul  2 06:17:49 CDT 2017
Git repo info: branch = master, commit = c05f1d24c6c6
Python 2.7.6, PyQt version 4.8.6
linux2
.saved: save-new-test.py
....setting leoID from os.getenv('USER'): 'vitalije'
....................test of at.printError: La Peña
..............s...........................
End of leoAtFile tests
..........s...sssss........................................
End of leoColor tests
.............................
End of leoCommands tests
........
End of leoConfig tests
......................................s..
.....ss.....................................s..s.................................................................................................sss....sss...
End of typing tests
.
End of leoEditCommands tests.
............
End of leoFileCommands tests.
.....
End of leoFind tests.
.............
End of leoFrame tests.
........................................................................
End of leoGlobals tests.
.sss
End of leoGui tests.
...............................................................................................................................
End of leoImport tests.
.......
End of leoKeys tests.
........................................
End of leoNodes tests.
....................ss..s..ss.....EEEEEE...........................................................
End of leoUndo tests.
..............................................................................................................................................................
End of plugins unit tests
.@test print redraw count: 653
..all unit tests done
..
======================================================================
ERROR: runTest (leo.core.leoTest.GeneralTestCase)
@test rst3Test @no-head

----------------------------------------------------------------------
Traceback (most recent call last):
  File "/media/vitalije/home11/vitalije/programi/leo-editor/trunk-git/leo/core/leoTest.py", line 211, in runTest
    builtins.execfile(scriptFile, d)
  File "/home/vitalije/.leo/scriptFile.py", line 6, in <module>
    rst3Test(c,p)
  File "<string>", line 14, in __init__
  File "<string>", line 54, in run
  File "<string>", line 43, in clean
AttributeError: 'NoneType' object has no attribute 'replace'

======================================================================
ERROR: runTest (leo.core.leoTest.GeneralTestCase)
@test rst3Test default

----------------------------------------------------------------------
Traceback (most recent call last):
  File "/media/vitalije/home11/vitalije/programi/leo-editor/trunk-git/leo/core/leoTest.py", line 211, in runTest
    builtins.execfile(scriptFile, d)
  File "/home/vitalije/.leo/scriptFile.py", line 6, in <module>
    rst3Test(c,p)
  File "<string>", line 14, in __init__
  File "<string>", line 54, in run
  File "<string>", line 43, in clean
AttributeError: 'NoneType' object has no attribute 'replace'

======================================================================
ERROR: runTest (leo.core.leoTest.GeneralTestCase)
@test rst3Test doc_only_mode (set in headline)

----------------------------------------------------------------------
Traceback (most recent call last):
  File "/media/vitalije/home11/vitalije/programi/leo-editor/trunk-git/leo/core/leoTest.py", line 211, in runTest
    builtins.execfile(scriptFile, d)
  File "/home/vitalije/.leo/scriptFile.py", line 7, in <module>
    rst3Test(c,p)
  File "<string>", line 14, in __init__
  File "<string>", line 54, in run
  File "<string>", line 43, in clean
AttributeError: 'NoneType' object has no attribute 'replace'

======================================================================
ERROR: runTest (leo.core.leoTest.GeneralTestCase)
@test rst3Test doc_only_mode (set in options doc part)

----------------------------------------------------------------------
Traceback (most recent call last):
  File "/media/vitalije/home11/vitalije/programi/leo-editor/trunk-git/leo/core/leoTest.py", line 211, in runTest
    builtins.execfile(scriptFile, d)
  File "/home/vitalije/.leo/scriptFile.py", line 7, in <module>
    rst3Test(c,p)
  File "<string>", line 14, in __init__
  File "<string>", line 54, in run
  File "<string>", line 43, in clean
AttributeError: 'NoneType' object has no attribute 'replace'

======================================================================
ERROR: runTest (leo.core.leoTest.GeneralTestCase)
@test rst3Test show_leo_directives=False

----------------------------------------------------------------------
Traceback (most recent call last):
  File "/media/vitalije/home11/vitalije/programi/leo-editor/trunk-git/leo/core/leoTest.py", line 211, in runTest
    builtins.execfile(scriptFile, d)
  File "/home/vitalije/.leo/scriptFile.py", line 7, in <module>
    rst3Test(c,p)
  File "<string>", line 14, in __init__
  File "<string>", line 54, in run
  File "<string>", line 43, in clean
AttributeError: 'NoneType' object has no attribute 'replace'

======================================================================
ERROR: runTest (leo.core.leoTest.GeneralTestCase)
@test rst3Test unicode characters

----------------------------------------------------------------------
Traceback (most recent call last):
  File "/media/vitalije/home11/vitalije/programi/leo-editor/trunk-git/leo/core/leoTest.py", line 211, in runTest
    builtins.execfile(scriptFile, d)
  File "/home/vitalije/.leo/scriptFile.py", line 10, in <module>
    rst3Test(c,p)
  File "<string>", line 14, in __init__
  File "<string>", line 54, in run
  File "<string>", line 43, in clean
AttributeError: 'NoneType' object has no attribute 'replace'

----------------------------------------------------------------------
Ran 905 tests in 98.595s

FAILED (errors=6, skipped=26)
`
I don't know why Terry's log reports 6 failures, 8 errors and 28 skipped. It seems he has more tests than I have or maybe I am not running them properly.
`
python launchLeo.py leo/test/unitTest.leo
`
and `Alt+4` after. Am I doing something wrong?

**Comment by edreamleo** [comment]
On Sun, Jul 2, 2017 at 8:23 AM, Виталије Милошевић <notifications@github.com
> wrote:

​[snip]​


> ERROR: runTest (leo.core.leoTest.GeneralTestCase)
> @test <https://github.com/test> rst3Test unicode characters
> ------------------------------
>
> Traceback (most recent call last):
> File "/media/vitalije/home11/vitalije/programi/leo-editor/
> trunk-git/leo/core/leoTest.py", line 211, in runTest
> builtins.execfile(scriptFile, d)
> File "/home/vitalije/.leo/scriptFile.py", line 10, in
> rst3Test(c,p)
> File "", line 14, in *init*
> File "", line 54, in run
> File "", line 43, in clean
> AttributeError: 'NoneType' object has no attribute 'replace'
> ------------------------------
>
> Ran 905 tests in 98.595s
>
> FAILED (errors=6, skipped=26)
> I don't know why Terry's log reports 6 failures, 8 errors and 28 skipped.
> It seems he has more tests than I have or maybe I am not running them
> properly.
> python launchLeo.py leo/test/unitTest.leo
> andAlt+4` after. Am I doing something wrong?
>

​I don't think so.  Probably subtle differences in platform.​​


​Thanks for this report.​


​Edward

**Comment by edreamleo** [comment]
On Sun, Jul 2, 2017 at 3:52 PM, Edward K. Ream <edreamleo@gmail.com> wrote:

​For everyone's information, the failure is happening in class rst3Test​,
defined *not *in leoTest.py as one would expect, but rather in
unitTest.leo, in a child of @common leoRst test code.

The pattern for all the failing tests is:

s = g.findTestScript(c,'@common leoRst test code',warn=False)
if s:
    exec(s)
    rst3Test(c,p)

There are two reasons I may have wanted to experiment with this pattern:

1. To make the test code in unitTest.leo independent of leoTest.py.

2. To create *standalone test files* so that tests can be run without Leo.

In retrospect, however, this pattern looks like a bad idea.

Since then, we now have the *prefix code pattern* that allows unit tests to
be rerun when code changes without reloading unitTest.leo.

In short, I'll move the rst3Test class into leoTest.leo.

Edward

**Comment by edreamleo** [comment]
On Sun, Jul 2, 2017 at 4:38 PM, Edward K. Ream <edreamleo@gmail.com> wrote:

​> ​
In short, I'll move the rst3Test class into leoTest.leo.

​Hmm. I just realized that there are other uses of the @common pattern in
unitTest.leo that don't come to grief. I'll attempt to fix the rst3Test
class first...

Edward

**Comment by vitalije** [comment]
Solved at 66b00d6

**Comment by vitalije** [comment]
it seems that we both were working on this. I have found that missing call_docutils=True was causing failure on my machine

**Comment by edreamleo** [comment]
On Sun, Jul 2, 2017 at 4:48 PM, Виталије Милошевић <notifications@github.com
> wrote:

> it seems that we both were working on this. I have found that missing
> call_docutils=True was causing failure on my machine
>
​Thanks for this.  I mistakenly overwrote your work.  I'm attempting to
revert my mistake.

Edward

**Comment by edreamleo** [comment]
On Sun, Jul 2, 2017 at 5:26 PM, Edward K. Ream <edreamleo@gmail.com> wrote:

>
>
> On Sun, Jul 2, 2017 at 4:48 PM, Виталије Милошевић <
> notifications@github.com> wrote:
>
>> it seems that we both were working on this. I have found that missing
>> call_docutils=True was causing failure on my machine
>>
> ​Thanks for this.  I mistakenly overwrote your work.  I'm attempting to
> revert my mistake.
>

​Rev 3d55e756e200​

​restores your commit.  I'll take a look at it, and possibly merge my
previous changes.

Edward

**Comment by edreamleo** [comment]
Rev c454507 merges my changes to the unit tests that use the rst3Test class, retaining Vitalije's changes.

I'll close this now.  Please reopen if there are any more failures.

**Comment by edreamleo** [comment]
Oops.  Terry still reports failures.  I'll reopen.

**Comment by vitalije** [comment]
I have solved for few more tests what was the problem. There are 3 remaining failures on my machine when runing with Python3. Alll three of them have the same line failing vr plugin can't be init-ed because QtSvg and QtWebKitWidgets are not present. I am using PyQt5.2.1 with Python3

**Comment by vitalije** [comment]
Yes, installing python3-pyqt5.qtsvg and python3-pyqt5.qtwebkit solved the remaining issues.
```
.@test print redraw count: 635
..all unit tests done
..
----------------------------------------------------------------------
Ran 906 tests in 92.831s

OK (skipped=42)
```
I will close this issue now hoping that @tbnorth will also have all tests pass.

**Comment by edreamleo** [comment]
On Mon, Jul 3, 2017 at 3:25 AM, Виталије Милошевић <notifications@github.com
> wrote:

> Yes, installing python3-pyqt5.qtsvg and python3-pyqt5.qtwebkit solved the
> remaining issues.
>
​Excellent sleuthing.

But don't we need to skip the affected unit tests, or alternatively, issue
a clear message about why the unit test failed.

Which 3 unit test fail without qtsvg and qtwebkit?

Edward

**Comment by edreamleo** [comment]
On Mon, Jul 3, 2017 at 5:15 AM, Edward K. Ream <edreamleo@gmail.com> wrote:

​> ​
Which 3 unit test fail without qtsvg and qtwebkit?

​To answer my own question, it looks like these 3 tests will fail if the vr
plugin can not be loaded:

@test helpForbindings
@test helpForFindCommands
@test helpForMinibuffer

Each fails the same way:

File ".../leo/plugins/qt_gui.py", line 1029, in put_help
    assert vr # For unit testing.


Vitalije, are these the three unit tests you referred to?

Edward

**Comment by vitalije** [comment]
I remember only one by name `test helpForFindCommands`, but the other two were very near.
There is assertion somewhere in 
`LeoPyRef.leo#Code-->Qt gui-->@file ../plugins/qt_gui.py-->class LeoQtGui(leoGui.LeoGui)-->qt_gui.put_help`
`assert vr if unittesting.`

All three tests were failing on this assertion. And that was because vr.init() returned False

**Comment by vitalije** [comment]
Yes those three tests.

**Comment by vitalije** [comment]
@edreamleo 
> But don't we need to skip the affected unit tests, or alternatively, issue
   a clear message about why the unit test failed.
Yes, that would be helpful. In a way the only thing needed is just to change assertion message.
`QtSvg and QtWebKitWidgets must be installed for this test to pass`

**Comment by edreamleo** [comment]
Rev 6c261e4 should cure failures on Terry's machine.  The unit tests succeed without testing all the help-for logic.

This could be called good enough.  The logic just needs to be tested for reversions, and the complete tests run on my machine.  The alternative would be to call self.skipTest('no vr plugin') *after* running the entire unit test!

But as I write this, I see that calling skipTest is more informative.  I'll do this next.

**Comment by edreamleo** [comment]
> But as I write this, I see that calling skipTest is more informative. I'll do this next.

Done at b536b45 in master.  Indeed, the call to skipTest is quite useful.  It will be called whenever running unit tests externally.

Unless Terry reports further problems, this will complete this project.

**Comment by tbnorth** [comment]
```
Leo 5.5, build 20170703073811, Mon Jul  3 07:38:11 CDT 2017
Git repo info: branch = master, commit = f0a76ae7b793
Python 2.7.12, PyQt version 5.5.1
linux2
```

I find running Leo for unit tests in its own VNC server instance is handy, because you can do other things without focus issues.  I'm getting these three errors:

```
Trying to create a QVariant instance of QMetaType::Void type, an invalid QVariant will be constructed instead
.
======================================================================
ERROR: runTest (leo.core.leoTest.GeneralTestCase)
@test x.makeShadowDirectory

----------------------------------------------------------------------
Traceback (most recent call last):
  File "/mnt/usr1/usr1/home/tbrown/t/Package/leo/git/leo-editor/leo/core/leoTest.py", line 211, in runTest
    builtins.execfile(scriptFile, d)
  File "/home/tbrown/.leo/scriptFile.py", line 31, in <module>
    deleteShadowDir(shadow_dir)
  File "/home/tbrown/.leo/scriptFile.py", line 20, in deleteShadowDir
    os.rmdir(shadow_dir)
OSError: [Errno 39] Directory not empty: '/home/tbrown/.leo/shadow'

======================================================================
FAIL: runTest (leo.core.leoTest.GeneralTestCase)
@test all commands have an event arg

----------------------------------------------------------------------
Traceback (most recent call last):
  File "/mnt/usr1/usr1/home/tbrown/t/Package/leo/git/leo-editor/leo/core/leoTest.py", line 211, in runTest
    builtins.execfile(scriptFile, d)
  File "/home/tbrown/.leo/scriptFile.py", line 26, in <module>
    assert arg0 in expected or arg1 in expected, message
AssertionError: 
no event arg for command quickmove_bookmark_to_first_child, func: <lambda>
args: ArgSpec(args=['e'], varargs=None, keywords=None, defaults=None)

======================================================================
FAIL: runTest (leo.core.leoTest.GeneralTestCase)
@test add/delete html comments

----------------------------------------------------------------------
Traceback (most recent call last):
  File "/mnt/usr1/usr1/home/tbrown/t/Package/leo/git/leo-editor/leo/core/leoTest.py", line 211, in runTest
    builtins.execfile(scriptFile, d)
  File "/home/tbrown/.leo/scriptFile.py", line 27, in <module>
    assert p.b == s,'fail5: s\n%s\nresult\n%s' % (repr(s),repr(p.b))
AssertionError: fail5: s
u'@language html\n<html>\n    text \n</html>\n'
result
u'@language html\n<html>\n   text \n</html>\n'

======================================================================
FAIL: runTest (leo.core.leoTest.EditBodyTestCase)
EditBodyTestCase: addComments
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/mnt/usr1/usr1/home/tbrown/t/Package/leo/git/leo-editor/leo/core/leoTest.py", line 105, in runTest
    self.editBody()
  File "/mnt/usr1/usr1/home/tbrown/t/Package/leo/git/leo-editor/leo/core/leoTest.py", line 85, in editBody
    compareHeadlines=False), '%s: before undo1' % commandName
AssertionError: addComments: before undo1

----------------------------------------------------------------------
Ran 904 tests in 68.079s

FAILED (failures=3, errors=1, skipped=44)
```

**Comment by xgid** [comment]
Hi there!

I'm also having 3 failed tests with current master, some of those seem to be different from Terry's ones:
```python
Trying to create a QVariant instance of QMetaType::Void type, an invalid QVariant will be constructed instead
.
======================================================================
FAIL: runTest (leo.core.leoTest.GeneralTestCase)
@test all commands have an event arg

----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\dev\Leo-5.4\leo\core\leoTest.py", line 209, in runTest
    exec(compile(script, scriptFile, 'exec'), d)
  File "C:\Users\sysadmin\.leo\scriptFile.py", line 26, in <module>
    assert arg0 in expected or arg1 in expected, message
AssertionError:
no event arg for command quickmove_bookmark_to_first_child, func: <lambda>
args: ArgSpec(args=['e'], varargs=None, keywords=None, defaults=None)

======================================================================
FAIL: runTest (leo.core.leoTest.GeneralTestCase)
@test g.app.config @buttons and @commands logic

----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\dev\Leo-5.4\leo\core\leoTest.py", line 209, in runTest
    exec(compile(script, scriptFile, 'exec'), d)
  File "C:\Users\sysadmin\.leo\scriptFile.py", line 26, in <module>
    assert fn in aList,'%s not in unitTestDict[%s]' % (fn,key)
AssertionError: myLeoSettings.leo not in unitTestDict[config.doCommands-file-names]

======================================================================
FAIL: runTest (leo.core.leoTest.GeneralTestCase)
@test unbound Alt-9 key is completely ignored

----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\dev\Leo-5.4\leo\core\leoTest.py", line 209, in runTest
    exec(compile(script, scriptFile, 'exec'), d)
  File "C:\Users\sysadmin\.leo\scriptFile.py", line 27, in <module>
    assert g.app.unitTestDict.get('handleUnboundChar-ignore-alt-or-ctrl')
AssertionError

----------------------------------------------------------------------
Ran 905 tests in 120.272s

FAILED (failures=3, skipped=11)
```

But most important, I have the Python process crashing at this point:
![python crash during unit tests](https://user-images.githubusercontent.com/12764559/27829637-e2246cbe-6098-11e7-8e17-f12b687c2643.png)

As it seems from the logs above that the tests had already completed, maybe the crash is just when finishing the test execution... but I'm not sure.

I can run VS to debug the crash... would you like any special info from the debugger? The call stack maybe?

**Comment by edreamleo** [comment]
On Tue, Jul 4, 2017 at 7:19 AM, XGiD <notifications@github.com> wrote:

> Hi there!
>
> I'm also having 3 failed tests with current master, some of those seem to
> be different from Terry's ones:
>
​Thanks for this report.

> I can run VS to debug the crash... would you like any special info from
> the debugger? The call stack maybe?
> ​
>
> ​Yes, send it on.

Edward

**Comment by xgid** [comment]
OK, this is what VS is showing:
![VS debugger stack](https://user-images.githubusercontent.com/12764559/27841424-7f9bb3ba-60d7-11e7-8b4a-075775c81e18.png)

The crash is systematic.

Any other info I may bring?

**Comment by edreamleo** [comment]
c8b888a completes this item.  All unit tests should pass (or be skipped) for both Terry and xgid.  Let me know if any fail.

I disabled the test: @test unbound Alt-9 key is completely ignored. It's a minor test that isn't worth the risk of hard crashes.

**Comment by xgid** [comment]
Good work Edward! All tests passing for me now with no crashes!!
```
Ran 914 tests in 122.572s

OK (skipped=12)
```

The only thing is that Leo is automatically closed after the last test completes, but maybe that's normal...?

**Comment by edreamleo** [comment]
On Wed, Jul 12, 2017 at 11:58 PM, XGiD <notifications@github.com> wrote:

> Good work Edward! All tests passing for me now with no crashes!!
>
​Thanks for the confirmation.

Terry, all good with you now?

Edward

**Comment by tbnorth** [comment]
On Thu, 13 Jul 2017 03:08:40 -0700
"Edward K. Ream" <notifications@github.com> wrote:

> On Wed, Jul 12, 2017 at 11:58 PM, XGiD <notifications@github.com>
> wrote:
> 
> > Good work Edward! All tests passing for me now with no crashes!!
> >
> ​Thanks for the confirmation.
> 
> Terry, all good with you now?

Yep, 

Ran 913 tests in 38.096s

OK (skipped=44)

which is much faster than usual, I'm sure it was over 100sec.  With my
personal settings enabled I get one fail in makeShadowDir(sp?), dir not
empty for some reason.  But all good with HOME=/tmp invocation.

Cheers -Terry

**Comment by edreamleo** [comment]
On Thu, Jul 13, 2017 at 7:35 AM, Terry Brown <notifications@github.com>
wrote:

> > Terry, all good with you now?
>
> Yep,
>
> Ran 913 tests in 38.096s
>
> OK (skipped=44)
>
> which is much faster than usual, I'm sure it was over 100sec. With my
> personal settings enabled I get one fail in makeShadowDir(sp?), dir not
> empty for some reason. But all good with HOME=/tmp invocation.
>

​@test x.makeShadowDirectory ​may depend on permissions.  It's a test on
code that is unlikely ever to change, so I am tempted to disable it. I'll
leave it for now, and close this item.

Edward

**Comment by edreamleo** [comment]
[#527](https://github.com/leo-editor/leo-editor/issues/527) is an Info item suggesting how to make unit tests independent of user settings, including settings in unitTest.leo.
